File: src/api/routes/eval.py
"""
Eval API routes.

API Layer Position:
  Frontend (/eval pages) → [routes] → eval package + storage + registry

Design decisions:
  - Long-running runs dispatched via FastAPI's BackgroundTasks; the
    POST returns 202 immediately with a run_id the client can poll.
  - RunRegistry tracks in-flight progress for the status endpoint;
    persisted runs live on disk via storage.save_run.
  - Pagination on results so a 200-question run doesn't ship 200KB+
    of generated text per page load.
  - 409 on eval-set version mismatch in compare so the UI can surface
    a clear "these runs aren't comparable" banner.
"""

from __future__ import annotations

import os
import subprocess
from datetime import datetime, timezone
from pathlib import Path

from fastapi import APIRouter, BackgroundTasks, HTTPException, Query, Request, status

from src.api.schemas.eval import (
    AggregatedMetricDTO,
    EvalResultDTO,
    RunDetailDTO,
    RunStatusDTO,
    RunSubmitRequest,
    RunSubmitResponse,
    RunSummaryDTO,
)
from src.api.services.eval_runs import RunRegistry
from src.eval.compare import compare_runs as _compare_runs_impl
from src.eval.config import load_config
from src.eval.runner import EvalRunner
from src.eval.schemas import CompareResult, EvalResult
from src.eval.storage import compute_run_id, list_runs, load_run

router = APIRouter(prefix="/api/eval", tags=["eval"])

# WHY: module-level attribute so monkeypatch.setattr("src.api.routes.eval.CONFIGS_DIR", ...)
# works in tests. Route handlers read this name from module globals at call time.
CONFIGS_DIR = Path("configs/eval")


# --------------------------------------------------------------------------- #
# Dependency helper                                                            #
# --------------------------------------------------------------------------- #

def _get_registry(request: Request) -> RunRegistry:
    """Extract the shared RunRegistry from app.state.

    PATTERN: Thin helper matching the get_backend pattern in dependencies.py.
    Routes call _get_registry(request) to stay testable and explicit.

    WHY lazy init: Starlette's TestClient does not run the lifespan when used
    outside a context manager (as in the test fixture). Lazy init guarantees
    a registry exists even when the lifespan startup hook hasn't fired —
    the registry is still correct because it's a plain in-memory dict.
    """
    if not hasattr(request.app.state, "run_registry"):
        # PATTERN: thread-safe because attribute assignment on a single object is
        # atomic in CPython; worst case two threads both create a registry and one
        # overwrites the other — acceptable for test scenarios.
        request.app.state.run_registry = RunRegistry()
    return request.app.state.run_registry


# --------------------------------------------------------------------------- #
# Background worker                                                            #
# --------------------------------------------------------------------------- #

def _run_eval_in_background(
    config_name: str,
    run_id: str,
    registry: RunRegistry,
) -> None:
    """Synchronous worker invoked via BackgroundTasks.

    WHY sync (not async): EvalRunner is CPU/IO-mixed and calls blocking LLM
    APIs. Sync BackgroundTasks workers are run in a threadpool by Starlette,
    keeping the event loop free. An async worker would block the loop.

    Pipeline position: DISPATCH — called once per POST /api/eval/run,
    runs the full EvalRunner lifecycle, then marks the run done/failed
    in the registry.
    """
    # WHY live import of storage: the tmp_eval_runs fixture reloads
    # src.eval.storage after setting EVAL_RUNS_DIR. Importing at call time
    # ensures we see the reloaded module attribute value.
    import src.eval.storage as _storage

    cfg_path = CONFIGS_DIR / f"{config_name}.yaml"
    cfg = load_config(cfg_path)

    # PATTERN: respect EVAL_LLM_OVERRIDE_DUMMY=1 — same logic as cli._cmd_run.
    # This makes the test harness fast (no real LLM calls).
    llm_override = None
    judge_llm_override = None
    if os.getenv("EVAL_LLM_OVERRIDE_DUMMY") == "1":
        from src.eval.cli import _DummyLLM
        dummy = _DummyLLM()
        llm_override = dummy
        judge_llm_override = dummy

    runner = EvalRunner(
        cfg,
        config_path=cfg_path,
        llm_override=llm_override,
        judge_llm_override=judge_llm_override,
        on_progress=lambda done, total: registry.update_progress(run_id, done),
        # WHY run_id_override: we pre-computed the run_id at submit time so the
        # registry could be populated before the run starts. Passing it here
        # ensures EvalRunner saves to the same directory the status endpoint expects.
        run_id_override=run_id,
    )

    try:
        runner.run()
        registry.mark_completed(run_id)
    except Exception as exc:
        registry.mark_failed(run_id, str(exc))


# --------------------------------------------------------------------------- #
# GET /api/eval/configs                                                        #
# --------------------------------------------------------------------------- #

@router.get(
    "/configs",
    response_model=list[str],
    summary="List available eval config names",
)
def list_configs() -> list[str]:
    """Return the names (without .yaml extension) of all configs in configs/eval/.

    The client uses this to populate the "New eval run" dropdown — no need
    to expose the full YAML content at this level.
    """
    # WHY: CONFIGS_DIR is read as a module global so monkeypatching works.
    if not CONFIGS_DIR.exists():
        return []
    return [p.stem for p in sorted(CONFIGS_DIR.glob("*.yaml"))]


# --------------------------------------------------------------------------- #
# POST /api/eval/run                                                           #
# --------------------------------------------------------------------------- #

@router.post(
    "/run",
    response_model=RunSubmitResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Submit an eval run (non-blocking)",
)
def submit_run(
    body: RunSubmitRequest,
    background_tasks: BackgroundTasks,
    request: Request,
) -> RunSubmitResponse:
    """Queue an eval run and return immediately with a run_id to poll.

    PATTERN: async job — the route validates the config exists, pre-computes
    the run_id (same algorithm as EvalRunner so they agree on the directory
    name), registers the run in the registry as "queued", dispatches via
    BackgroundTasks, then returns 202. The client polls /runs/{run_id}/status.

    WHY pre-compute run_id: the registry must track the run BEFORE it
    starts, so the status endpoint can return "queued" immediately after
    submission. EvalRunner accepts run_id_override to use the same id.
    """
    config_name = body.config_name
    cfg_path = CONFIGS_DIR / f"{config_name}.yaml"

    # 404 if the config file doesn't exist.
    if not cfg_path.exists():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Config '{config_name}' not found in {CONFIGS_DIR}.",
        )

    # Pre-compute run_id using the same algorithm as EvalRunner.run().
    started_at = datetime.now(timezone.utc)
    try:
        git_sha = subprocess.check_output(
            ["git", "rev-parse", "HEAD"], text=True
        ).strip()
    except Exception:
        git_sha = "unknown"

    run_id = compute_run_id(config_name, started_at, git_sha)

    # Register before dispatch so status can return "queued" immediately.
    registry = _get_registry(request)
    # WHY n_total=0: we don't know question count until the runner loads datasets.
    # update_progress transitions the entry to "running" on first call.
    registry.register(run_id, n_total=0)

    # Dispatch the synchronous worker via BackgroundTasks (runs in threadpool).
    background_tasks.add_task(_run_eval_in_background, config_name, run_id, registry)

    return RunSubmitResponse(run_id=run_id, status="queued")


# --------------------------------------------------------------------------- #
# GET /api/eval/runs                                                           #
# --------------------------------------------------------------------------- #

@router.get(
    "/runs",
    response_model=list[RunSummaryDTO],
    summary="List all completed eval runs",
)
def list_eval_runs() -> list[RunSummaryDTO]:
    """Return all persisted runs sorted by started_at descending.

    WHY disk-only: completed runs are on disk; in-flight runs lack aggregated
    metrics so they aren't useful in the list view. The status endpoint covers
    in-flight monitoring.
    """
    # WHY live import of list_runs: called via the function (which reads
    # EVAL_RUNS_DIR from the module global at call time), so the reloaded
    # module attribute is always used correctly.
    runs = list_runs()
    result: list[RunSummaryDTO] = []
    for meta in runs:
        # Compute headline metric: recall_at_5 mean if available, else None.
        headline: float | None = None
        try:
            run_data = load_run(meta.run_id)
            aggregated = run_data["aggregated"]
            for agg in aggregated:
                if agg.metric_name == "recall_at_5" and agg.dataset is not None:
                    headline = agg.mean
                    break
        except Exception:
            # TRADE-OFF: if a run is partially written, skip its headline metric.
            pass
        result.append(
            RunSummaryDTO(
                run_id=meta.run_id,
                config_name=meta.config_name,
                started_at=meta.started_at,
                finished_at=meta.finished_at,
                n_questions=meta.n_questions,
                n_errors=meta.n_errors,
                headline_metric=headline,
            )
        )
    return result


# --------------------------------------------------------------------------- #
# GET /api/eval/runs/{run_id}                                                  #
# --------------------------------------------------------------------------- #

@router.get(
    "/runs/{run_id}",
    response_model=RunDetailDTO,
    summary="Get full detail for one eval run",
)
def get_run(run_id: str) -> RunDetailDTO:
    """Return metadata + aggregated metrics + cost for a single run.

    Results are paginated separately — this response is bounded regardless
    of how many questions the run evaluated.
    """
    try:
        run_data = load_run(run_id)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run '{run_id}' not found.",
        )

    meta = run_data["metadata"]
    aggregated = run_data["aggregated"]
    cost = run_data["cost"]
    results = run_data["results"]

    agg_dtos = [
        AggregatedMetricDTO(
            metric_name=a.metric_name,
            dataset=a.dataset,
            mean=a.mean,
            ci_low=a.ci_low,
            ci_high=a.ci_high,
            n=a.n,
        )
        for a in aggregated
    ]

    return RunDetailDTO(
        metadata=meta,
        aggregated=agg_dtos,
        cost=cost,
        n_results=len(results),
    )


# --------------------------------------------------------------------------- #
# GET /api/eval/runs/{run_id}/results                                          #
# --------------------------------------------------------------------------- #

@router.get(
    "/runs/{run_id}/results",
    summary="Paginated per-question results for a run",
)
def get_run_results(
    run_id: str,
    page: int = Query(default=1, ge=1, description="1-indexed page number"),
    page_size: int = Query(default=50, ge=1, le=200, description="Items per page (max 200)"),
) -> dict:
    """Return a paginated slice of per-question results.

    WHY paginated: a 200-question run has ~200KB of generated text + metrics.
    Streaming the whole payload on every page load is wasteful; pagination
    lets the UI load one screenful at a time.

    Returns:
        Dict with keys: items, page, page_size, total.
    """
    try:
        run_data = load_run(run_id)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run '{run_id}' not found.",
        )

    results: list[EvalResult] = run_data["results"]
    total = len(results)

    # WHY 0-indexed slice: page=1 → items[0:page_size], page=2 → items[page_size:2*page_size].
    start = (page - 1) * page_size
    end = start + page_size
    page_items = results[start:end]

    dtos = [
        EvalResultDTO(
            question_id=r.question_id,
            dataset=r.dataset,
            generated_answer=r.generated_answer,
            metrics=r.metrics,
            error=r.error,
        )
        for r in page_items
    ]

    return {"items": [d.model_dump() for d in dtos], "page": page, "page_size": page_size, "total": total}


# --------------------------------------------------------------------------- #
# GET /api/eval/runs/{run_id}/results/{question_id}                            #
# --------------------------------------------------------------------------- #

@router.get(
    "/runs/{run_id}/results/{question_id}",
    response_model=EvalResult,
    summary="Get full EvalResult for one question",
)
def get_question_result(run_id: str, question_id: str) -> EvalResult:
    """Return the full EvalResult (including retrieved_chunks, metric_details).

    WHY separate endpoint: the list view uses EvalResultDTO (slim), which omits
    the large fields. This endpoint returns the full schema for the detail drawer.
    """
    try:
        run_data = load_run(run_id)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run '{run_id}' not found.",
        )

    results: list[EvalResult] = run_data["results"]
    for r in results:
        if r.question_id == question_id:
            return r

    raise HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail=f"Question '{question_id}' not found in run '{run_id}'.",
    )


# --------------------------------------------------------------------------- #
# GET /api/eval/runs/{run_id}/status                                           #
# --------------------------------------------------------------------------- #

@router.get(
    "/runs/{run_id}/status",
    response_model=RunStatusDTO,
    summary="Poll the status of an in-progress or completed eval run",
)
def get_run_status(run_id: str, request: Request) -> RunStatusDTO:
    """Return the current status and progress for a run.

    WHY two-phase lookup:
      1. Check in-memory registry (covers queued/running/recently-completed).
      2. If not in registry, check disk (covers runs from previous server
         restarts that the registry has evicted).
      3. 404 if neither source has the run.
    """
    registry = _get_registry(request)
    entry = registry.get(run_id)

    if entry is not None:
        progress = (
            (entry.n_completed / entry.n_total)
            if entry.n_total > 0 and entry.status == "completed"
            else (1.0 if entry.status == "completed" else 0.0)
        )
        return RunStatusDTO(
            run_id=run_id,
            status=entry.status,
            progress=progress,
            n_completed=entry.n_completed,
            n_total=entry.n_total,
            error_message=entry.error_message,
        )

    # Fallback: check disk for runs not in the registry (e.g. post-restart).
    try:
        run_data = load_run(run_id)
        meta = run_data["metadata"]
        return RunStatusDTO(
            run_id=run_id,
            status="completed",
            progress=1.0,
            n_completed=meta.n_questions,
            n_total=meta.n_questions,
            error_message=None,
        )
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run '{run_id}' not found.",
        )


# --------------------------------------------------------------------------- #
# GET /api/eval/compare                                                        #
# --------------------------------------------------------------------------- #

@router.get(
    "/compare",
    response_model=CompareResult,
    summary="Compare two eval runs (delta metrics + per-question diff)",
)
def compare_runs(
    a: str = Query(..., description="run_id of run A"),
    b: str = Query(..., description="run_id of run B"),
) -> CompareResult:
    """Compute metric deltas and per-question diffs between two runs.

    TRADE-OFF: 409 on eval-set version mismatch rather than silently computing
    a misleading comparison. Different eval-set versions mean different questions,
    so per-question deltas are meaningless. Surfaces this clearly to the UI.
    """
    try:
        return _compare_runs_impl(a, b)
    except FileNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )
    except ValueError as exc:
        # WHY 409 (Conflict): the comparison is a logical conflict — the two
        # runs are not comparable because they evaluated different question sets.
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(exc),
        )

File: frontend/src/api/eval.ts
/**
 * Eval API client + TanStack Query hooks.
 *
 * API Layer Position:
 *   React components → [hooks] → fetch → /api/eval/*
 *
 * Design notes:
 *   - Field names match the backend's snake_case JSON wire format, consistent
 *     with the convention in types.ts (e.g. doc_id, chunk_id, created_at).
 *   - Uses the same `request<T>` helper pattern as client.ts — a thin wrapper
 *     around fetch that throws on non-2xx with the backend's `detail` message.
 *   - useRunStatus polls every 1s by default; pass `refetchInterval: false`
 *     once a run hits "completed" / "failed" to stop polling.
 *   - Submitting a run invalidates the runs list so the new run shows
 *     up immediately.
 */

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

// ---------------------------------------------------------------------------
// Types — mirror the backend DTOs from src/api/schemas/eval.py and
// src/eval/schemas.py. Field names stay snake_case to match JSON wire format,
// consistent with how types.ts handles doc_id, chunk_id, created_at, etc.
// ---------------------------------------------------------------------------

export interface RunMetadata {
  run_id: string;
  config_name: string;
  config_path: string;
  git_sha: string;
  started_at: string; // ISO timestamp
  finished_at: string;
  env_hash: string;
  eval_set_versions: Record<string, string>;
  n_questions: number;
  n_errors: number;
  warnings: string[];
}

export interface RunSummary {
  run_id: string;
  config_name: string;
  started_at: string;
  finished_at: string;
  n_questions: number;
  n_errors: number;
  headline_metric: number | null;
}

export interface AggregatedMetric {
  metric_name: string;
  dataset: string | null;
  mean: number;
  ci_low: number;
  ci_high: number;
  n: number;
}

export interface RunDetail {
  metadata: RunMetadata;
  aggregated: AggregatedMetric[];
  cost: Record<string, number>;
  n_results: number;
}

export interface EvalResultRow {
  question_id: string;
  dataset: string;
  generated_answer: string;
  metrics: Record<string, number>;
  error: string | null;
}

export interface PageResults {
  items: EvalResultRow[];
  page: number;
  page_size: number;
  total: number;
}

export interface RunStatus {
  run_id: string;
  status: "queued" | "running" | "completed" | "failed";
  progress: number;
  n_completed: number;
  n_total: number;
  error_message: string | null;
}

export interface MetricDelta {
  metric_name: string;
  dataset: string | null;
  a_mean: number;
  a_ci: [number, number];
  b_mean: number;
  b_ci: [number, number];
  delta: number;
  p_value: number;
  significant: boolean;
}

export interface CompareResult {
  run_a: RunMetadata;
  run_b: RunMetadata;
  deltas: MetricDelta[];
  per_question_diff: Array<Record<string, unknown>>;
}

// ---------------------------------------------------------------------------
// Fetch helpers — mirror the request<T> helper from client.ts: same base URL
// env var, same error contract (throws Error with backend's detail message).
// WHY: a separate request helper here avoids coupling this module to client.ts
//      while staying consistent — easier to read in isolation.
// ---------------------------------------------------------------------------

const BASE_URL = import.meta.env.VITE_API_URL ?? "";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE_URL}${path}`, {
    ...init,
    headers: { ...init?.headers },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail ?? `Request failed: ${res.status}`);
  }
  return res.json();
}

// ---------------------------------------------------------------------------
// API functions — one per backend endpoint.
// ---------------------------------------------------------------------------

export async function listConfigs(): Promise<string[]> {
  return request<string[]>("/api/eval/configs");
}

export async function listRuns(): Promise<RunSummary[]> {
  return request<RunSummary[]>("/api/eval/runs");
}

export async function getRun(runId: string): Promise<RunDetail> {
  return request<RunDetail>(`/api/eval/runs/${encodeURIComponent(runId)}`);
}

export async function getRunResults(
  runId: string,
  page = 1,
  pageSize = 50,
): Promise<PageResults> {
  const qs = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });
  return request<PageResults>(
    `/api/eval/runs/${encodeURIComponent(runId)}/results?${qs}`,
  );
}

export async function getRunResult(
  runId: string,
  questionId: string,
): Promise<unknown> {
  return request(
    `/api/eval/runs/${encodeURIComponent(runId)}/results/${encodeURIComponent(questionId)}`,
  );
}

export async function getRunStatus(runId: string): Promise<RunStatus> {
  return request<RunStatus>(
    `/api/eval/runs/${encodeURIComponent(runId)}/status`,
  );
}

export async function submitRun(
  configName: string,
): Promise<{ run_id: string; status: RunStatus["status"] }> {
  return request("/api/eval/run", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ config_name: configName }),
  });
}

export async function compareRuns(
  idA: string,
  idB: string,
): Promise<CompareResult> {
  const qs = new URLSearchParams({ a: idA, b: idB });
  return request<CompareResult>(`/api/eval/compare?${qs}`);
}

// ---------------------------------------------------------------------------
// TanStack Query hooks — one per query/mutation, matching the style in
// use-documents.ts and use-conversations.ts (named exports, no default).
// Query keys follow the existing pattern: ["noun"] or ["noun", param, ...].
// ---------------------------------------------------------------------------

export function useConfigs() {
  return useQuery({ queryKey: ["eval-configs"], queryFn: listConfigs });
}

export function useRunsList() {
  return useQuery({ queryKey: ["eval-runs"], queryFn: listRuns });
}

export function useRun(runId: string | undefined) {
  return useQuery({
    queryKey: ["eval-run", runId],
    queryFn: () => getRun(runId!),
    enabled: !!runId,
  });
}

export function useRunResults(
  runId: string | undefined,
  page = 1,
  pageSize = 50,
) {
  return useQuery({
    queryKey: ["eval-run-results", runId, page, pageSize],
    queryFn: () => getRunResults(runId!, page, pageSize),
    enabled: !!runId,
  });
}

// PATTERN: refetchInterval drives live progress updates for in-flight runs.
//          The caller should pass `refetchInterval: false` once status reaches
//          "completed" or "failed" to stop polling and save network calls.
export function useRunStatus(
  runId: string | undefined,
  opts?: { refetchInterval?: number | false },
) {
  return useQuery({
    queryKey: ["eval-run-status", runId],
    queryFn: () => getRunStatus(runId!),
    enabled: !!runId,
    refetchInterval: opts?.refetchInterval ?? 1000,
  });
}

export function useCompareRuns(
  idA: string | undefined,
  idB: string | undefined,
) {
  return useQuery({
    queryKey: ["eval-compare", idA, idB],
    queryFn: () => compareRuns(idA!, idB!),
    enabled: !!idA && !!idB,
  });
}

export function useSubmitRun() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (configName: string) => submitRun(configName),
    // WHY: invalidate the runs list so the newly queued run appears
    //      immediately in the UI without a manual page refresh.
    onSuccess: () => qc.invalidateQueries({ queryKey: ["eval-runs"] }),
  });
}
